src/coreclr/jit/layout.cpp

```diff
@@ -746,9 +810,92 @@ bool ClassLayout::AreCompatible(const ClassLayout* layout1, const ClassLayout* l
 //
 bool ClassLayout::CanAssignFrom(const ClassLayout* layout)
 {
-    // Currently this is the same as compatability
+    if (this == layout)
+    {
+        return true;
+    }
+
+    // Do the normal compatibility check first, when possible to do so.
+    //
+    if ((IsCustomLayout() == layout->IsCustomLayout()) || (!HasGCPtr() && !layout->HasGCPtr()))
+    {
+        const bool areCompatible = AreCompatible(this, layout);
+
+        if (areCompatible)
+        {
+            return true;
+        }
+    }
```

Comment:
Doesn't the logic below subsume this check?

Reply:
I wanted to make this a superset of what `AreCompatible` does but that ended up being a bit clunky since I need to guard the call to avoid asserts.

I can either just give up on that idea or else give `AreCompatible` and optional early out to avoid asserting and just return false.

Reply:
What is the assert?

I would probably vote to just take the small amount of duplication. Seems simpler to reason about the operation that way.

Reply:
I think that assert is just outdated, it should have been removed as part of #112064 since we can now have layouts without class handles but with GC pointers

Reply:
In that case I can just remove the guarding and call it unconditionally.